### PR TITLE
Refactor Experience Index organization and Test Database Schema

### PR DESCRIPTION
**Pull request by lebaudantoine**
## Purpose

Test Experience Index Schema

## Proposal

Added unit tests on the two main database schema, Experience and Relation.

- [ ] Discuss timezone while storing timestamps in the Database
- [ ] Discuss Json validation

### CODE REVIEW

File: src/api/core/warren/tests/xi/schema/test_experience.py [hunk 1]
```diff
@@ -0,0 +1,807 @@
...
+    assert len(experiences) == 1
+
+    # Attempt creating another experience with the same ID
+    with pytest.raises(IntegrityError) as exc_info:
+        ExperienceFactory.create_sync(id=duplicated_id)
+
+    # Experience's creation should fail due to ID uniqueness
+    assert 'duplicate key value violates unique constraint "experience_pkey"' in str(
+        exc_info.value
+    )
```
Comment (jmaupetit) [inline src/api/core/warren/tests/xi/schema/test_experience.py:244]: ditto

Comment (jmaupetit) [reply src/api/core/warren/tests/xi/schema/test_experience.py:244]: (and also for following tests)

File: src/api/core/warren/tests/xi/schema/test_experience.py [hunk 2]
```diff
@@ -0,0 +1,807 @@
...
+    experiences = db_session.exec(select(Experience)).all()
+    assert len(experiences) == number_experiences + 1
+
+    db_session.refresh(expected_experience)
+
+    # Assert the retrieved experience matches the expected one
+    saved_experience = db_session.exec(
+        select(Experience).where(Experience.iri == iri)
+    ).one()
+    assert expected_experience == saved_experience
```
Comment (jmaupetit) [inline src/api/core/warren/tests/xi/schema/test_experience.py:404]: I don't get the purpose of this test. I have the feeling that you are testing the ORM or the database capability to execute a request.

File: src/api/core/warren/tests/xi/schema/test_experience.py [hunk 3]
```diff
@@ -0,0 +1,807 @@
...
+    with pytest.raises(DataError) as exc_info:
+        db_session.add(experience)
+        db_session.commit()
+
+    # Experience's creation should fail due to the max-length constraint
+    assert "value too long for type character" in str(exc_info.value)
+
+
+def test_experience_read(db_session: Session):
+    """Test retrieving detailed some experiences."""
```
Comment (jmaupetit) [inline src/api/core/warren/tests/xi/schema/test_experience.py:375]: Need rephrasing :sweat_smile:

File: src/api/core/warren/tests/xi/schema/test_experience.py [hunk 4]
```diff
@@ -0,0 +1,807 @@
...
+
+    # Create a given experience in the database
+    expected_experience = ExperienceFactory.create_sync()
+
+    # Attempt to read an experience by its ID
+    experience = db_session.get(Experience, expected_experience.id)
+
+    # Assert the retrieved experience matches the expected one
+    db_session.refresh(expected_experience)
+    assert expected_experience == experience
```
Comment (jmaupetit) [inline src/api/core/warren/tests/xi/schema/test_experience.py:441]: Are you testing your factory or Polyfactory itself?

File: src/api/core/warren/tests/xi/schema/test_experience.py [hunk 5]
```diff
@@ -0,0 +1,807 @@
...
+    # Attempt updating the second experience with first experience's ID
+    second_experience.id = first_experience.id
+    with pytest.raises(IntegrityError) as exc_info:
+        db_session.add(second_experience)
+        db_session.commit()
+
+    # Experience's update should fail due to ID uniqueness
+    assert 'duplicate key value violates unique constraint "experience_pkey"' in str(
+        exc_info.value
+    )
```
Comment (jmaupetit) [inline src/api/core/warren/tests/xi/schema/test_experience.py:593]: This constraint as already been tested.

File: src/api/core/warren/tests/xi/schema/test_experience.py [hunk 6]
```diff
@@ -0,0 +1,807 @@
...
+    # Assert the database contains one experience
+    experiences = db_session.exec(select(Experience)).all()
+    assert len(experiences) == 1
+
+    db_session.delete(experience)
+    db_session.commit()
+
+    # Assert the database is empty
+    experiences = db_session.exec(select(Experience)).all()
+    assert len(experiences) == 0
```
Comment (jmaupetit) [inline src/api/core/warren/tests/xi/schema/test_experience.py:780]: This is testing SQLAlchemy.

File: src/api/core/warren/tests/xi/schema/test_experience.py (hunk 5 above)
Comment (lebaudantoine) [reply src/api/core/warren/tests/xi/schema/test_experience.py:593]: I have tested this constraint while creating an experience. Here I test the update. Did I miss smth?

File: src/api/core/warren/tests/xi/schema/test_experience.py [hunk 7]
```diff
@@ -0,0 +1,807 @@
...
+    assert len(experiences) == 1
+
+    # Attempt creating another experience with the same IRI
+    with pytest.raises(IntegrityError) as exc_info:
+        ExperienceFactory.create_sync(iri=duplicated_iri)
+
+    # Experience's creation should fail due to IRI uniqueness
+    assert 'duplicate key value violates unique constraint "experience_iri_key"' in str(
+        exc_info.value
+    )
```
[the remark this answers is not in the document]
Comment (lebaudantoine) [reply src/api/core/warren/tests/xi/schema/test_experience.py:222]: More condensed, If you prefer, will do !

File: src/api/core/warren/tests/xi/schema/test_experience.py [hunk 8]
```diff
@@ -0,0 +1,807 @@
...
+    with pytest.raises(IntegrityError) as exc_info:
+        ExperienceFactory.create_sync(created_at=creation_date, updated_at=update_date)
+
+    # Experience's creation should fail due to the pre-creation 'update_date'
+    assert '"experience" violates check constraint "pre-creation-update"' in str(
+        exc_info.value
+    )
+
+
+# todo - discuss utc conversion logics for timestamps
```
[the remark this answers is not in the document]
Comment (lebaudantoine) [reply src/api/core/warren/tests/xi/schema/test_experience.py:176]: We discuss it on the Friday's stand-up. Let's catch up on it monday.

File: src/api/core/warren/tests/xi/schema/test_experience.py [hunk 9]
```diff
@@ -0,0 +1,807 @@
...
+            created_at=time,
+            updated_at=time,
+            relations_source=[],
+            relations_target=[],
+            **experience_data,
+        )
+
+        # Assert the database contains n experiences
+        experiences = db_session.exec(select(Experience)).all()
+        assert len(experiences) == n
```
[the remark this answers is not in the document]
Comment (lebaudantoine) [reply src/api/core/warren/tests/xi/schema/test_experience.py:86]: More factorized, yes ! Less explicit tho. Will do if you prefer this syntax.

File: src/api/core/warren/tests/xi/schema/test_experience.py (hunk 6 above)
Comment (lebaudantoine) [reply src/api/core/warren/tests/xi/schema/test_experience.py:780]: I am testing CRUD operation on my model, it feels weird to ignore such case, let's discuss it.
Agree that it's a pretty simple and straight-forward test.

File: src/api/core/warren/tests/xi/schema/test_experience.py (hunk 4 above)
Comment (lebaudantoine) [reply src/api/core/warren/tests/xi/schema/test_experience.py:441]: I was a bit lazy, but I can refactor this test to improve a bit its logic. I wanted to test, when reading an experience by its id, I would get the right data.

File: src/api/core/warren/tests/xi/schema/test_experience.py (hunk 2 above)
Comment (lebaudantoine) [reply src/api/core/warren/tests/xi/schema/test_experience.py:404]: I test the ORM or database capability to execute a request with my model. If you think it's vague or less a unit test but more an integration test, I can remove it.

File: src/api/core/warren/tests/xi/schema/test_experience.py (hunk 3 above)
Comment (lebaudantoine) [reply src/api/core/warren/tests/xi/schema/test_experience.py:375]:
```suggestion
    """Test retrieving details from some experiences."""
```

File: src/api/core/warren/tests/xi/schema/test_experience.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lebaudantoine) [reply src/api/core/warren/tests/xi/schema/test_experience.py:86]: Actually it doesn't work with `db_session.exec`, but with `db_session.query` which is deprecated

[the remark this answers is not in the document]
Comment (lebaudantoine) [reply src/api/core/warren/tests/xi/schema/test_experience.py:86]: I would keep it that way, as it's not a big deal.

File: src/api/core/warren/migrations/versions/77a0f0fbb8ab_fix_duration_constraint.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    op.drop_constraint("positive-duration", "experience", type_="check")
+    op.create_check_constraint("positive-duration", "experience", "duration > 0")
+    # ### end Alembic commands ###
+
+
+def downgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.drop_constraint("positive-duration", "experience", type_="check")
+    op.create_check_constraint("positive-duration", "experience", "duration >= 0")
+    # ### end Alembic commands ###
```
[the remark this answers is not in the document]
Comment (lebaudantoine) [reply src/api/core/warren/migrations/versions/77a0f0fbb8ab_fix_duration_constraint.py:33]: If I do so we would need to drop the prep rod db and run the migration again, let's merge it as it is and then, squash all migration before the release

File: src/api/core/warren/tests/xi/schema/test_experience.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (lebaudantoine) [reply src/api/core/warren/tests/xi/schema/test_experience.py]: We can improve this part in a dedicated PR, I can open an issue in the repo to keep track of this part.

**Review by jmaupetit — APPROVED** [review]

**Pull request merged**